app/health/checker.go

```diff
@@ -91,6 +96,33 @@ func (c *Checker) scrape() error {
 		return errors.Wrap(err, "gather metrics")
 	}
 
+	// Checking metrics with high cardinality.
+	var gatherAgain bool
+	for _, fams := range metrics {
+		if fams.GetName() == "app_health_metrics_high_cardinality" {
+			continue
+		}
+
+		var maxLabelsCount int
+		for _, fam := range fams.GetMetric() {
+			labelsCount := len(fam.GetLabel())
+			if labelsCount > maxLabelsCount {
+				maxLabelsCount = labelsCount
+			}
+		}
+		if maxLabelsCount > labelsCardinalityThreshold*c.numValidators {
+			highCardinalityGauge.WithLabelValues(fams.GetName()).Set(float64(maxLabelsCount))
+			gatherAgain = true
+		}
+	}
+
+	if gatherAgain {
+		metrics, err = c.gatherer.Gather()
+		if err != nil {
+			return errors.Wrap(err, "gather metrics")
+		}
+	}
```

Comment:
What is the reason behind this?

Reply:
Line 114: https://github.com/ObolNetwork/charon/pull/2967/files#diff-16436fa4ead5f4fac86aa0910e7dcef9ae20ab0c2fbc25365de419227dbaac2fR114
updates one gauge, we need to re-gather to have new set of metrics including the updated one.

Reply:
Shouldn't we run then the same for loop for the families again on the re-gathered metrics?

Reply:
This will be endless loop. We make assumption that some metrics may change in between of `Gather()` calls, but this ensures, the updated gauge (line 114) used for health checks is certainly included.